crates/ethereum/src/lib.rs

```diff
@@ -1,111 +1,434 @@
 }
 
-impl TryFrom<&ethers::types::Log> for BlockOrigin {
-    type Error = anyhow::Error;
+impl EthereumClient {
+    async fn get_block_number(&self) -> anyhow::Result<U256> {
+        let req = serde_json::json!({
+            "jsonrpc": "2.0",
+            "method": "eth_blockNumber",
+            "params": [],
+            "id" :0
+        });
+        let res: serde_json::Value = self
+            .http
+            .post(self.url.as_str())
+            .json(&req)
+            .send()
+            .await?
+            .json()
+            .await?;
+        let block_number = res["result"]
+            .as_str()
+            .and_then(|txt| U256::from_str_radix(txt, 16).ok())
+            .ok_or(anyhow::anyhow!("Failed to get block number"))?;
+        Ok(block_number)
+    }
 
-    fn try_from(log: &ethers::types::Log) -> Result<Self, Self::Error> {
-        let hash = log.block_hash.context("missing block hash")?;
-        let hash = EthereumBlockHash(hash);
-        let number = log.block_number.context("missing block number")?.as_u64();
-        let number = EthereumBlockNumber(number);
-        Ok(Self { hash, number })
+    pub async fn gas_price(&self) -> anyhow::Result<U256> {
+        let req = serde_json::json!({
+            "jsonrpc": "2.0",
+            "method": "eth_gasPrice",
+            "params": [],
+            "id" :0
+        });
+        let res: serde_json::Value = self
+            .http
+            .post(self.url.as_str())
+            .json(&req)
+            .send()
+            .await?
+            .json()
+            .await?;
+        let block_number = res["result"]
+            .as_str()
+            .and_then(|txt| U256::from_str_radix(txt, 16).ok())
+            .ok_or(anyhow::anyhow!("Failed to get gas price"))?;
+        Ok(block_number)
+    }
+
+    pub async fn chain_id(&self) -> anyhow::Result<EthereumChain> {
+        let req = serde_json::json!({
+            "jsonrpc": "2.0",
+            "method": "eth_chainId",
+            "params": [],
+            "id" :0
+        });
+        let res: serde_json::Value = self
+            .http
+            .post(self.url.as_str())
+            .json(&req)
+            .send()
+            .await?
+            .json()
+            .await?;
+        let id = res["result"]
+            .as_str()
+            .and_then(|txt| U256::from_str_radix(txt, 16).ok())
+            .ok_or(anyhow::anyhow!("Failed to get chain id"))?;
+        Ok(match id {
+            x if x == U256::from(1u32) => EthereumChain::Mainnet,
+            x if x == U256::from(5u32) => EthereumChain::Goerli,
+            x => EthereumChain::Other(x),
+        })
     }
 }
 
-impl TryFrom<&ethers::types::Log> for TransactionOrigin {
-    type Error = anyhow::Error;
+impl StarknetEthereumClient {
+    pub async fn get_starknet_state(&self) -> anyhow::Result<L1StateUpdate> {
+        let eth_block_number = self.eth.get_block_number().await?;
```

Comment:
Wait, but where is the race exactly? 

The `eth_call` being called against specific block number (which is believed to be the latest at the moment of the call), and the result is returned against the state of the block at the time of the call. If reorg happens during the call, it will be picked up on the next iteration. If block changes due to the reorg, it's number remains (at least I expect it to, as I'm passing it as a parameter explicitly).

I don't see how block hash solves the problem. It will actually require more housekeeping to make sure block hash and number are consistent.

Reply:
Done